### PR TITLE
Support predicate expressions

### PR DESCRIPTION
**Pull request by jievince**
https://github.com/vesoft-inc/nebula-graph/pull/467 depends on this pr.
doc: https://neo4j.com/docs/cypher-manual/current/functions/predicate/

### CODE REVIEW

**Comment by Shylock-Hg** [comment]
Why not put them into functions?

**Comment by jievince** [comment]
> Why not put them into functions?
To users, they are functions, but in implementation, they cannot be implemented as normal functions, because
1. Arguments of predicate functions are special, such as `all(var in list where filter)`, there are conflicts in parser.yy if define it into function_call_exprssion;
2. neo4j also implements them as expressions.

File: src/common/expression/PredicateExpression.cpp [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+        result_ = Value::kNullValue;
+        return result_;
+    }
+    if (!listVal.isList()) {
+        result_ = Value::kNullBadType;
+        return result_;
+    }
+    auto& list = listVal.getList();
+
+    if (name == "all") {
```
[the remark this answers is not in the document]
Comment (jievince) [reply src/common/expression/PredicateExpression.cpp:25]: Sorry,  I can't understand. 
1. Make the variable `name` to be a constexpr? 
2. How can enum variables be strings?

**Review by nevermore3 — APPROVED** [review]

**Review by CPWstatic — APPROVED** [review]
LGTM.

File: src/common/expression/PredicateExpression.cpp [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+    std::string buf;
+    buf.reserve(256);
+
+    buf += *name_;
+    buf += "(";
+    buf += *innerVar_;
+    buf += " IN ";
+    buf += collection_->toString();
+    buf += " WHERE ";
+    buf += filter_->toString();
```
Comment (yixinglu) [inline src/common/expression/PredicateExpression.cpp:184]: Does this expression always have `WHERE` clause ?

Comment (jievince) [reply src/common/expression/PredicateExpression.cpp:184]: Yes, it always has `WHERE`

**Review by yixinglu — APPROVED** [review]

**Pull request merged**